File: tests/benchmarks/http/http_server.go
// http_server.go - Go HTTP server (net/http) for benchmark.
//
// Usage:
//   go build -o http_server_go http_server.go && ./http_server_go [port]
//
// Default port: 8080

//go:build ignore

package main

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"runtime"
)

var jsonResp = []byte(`{"message":"Hello, World!"}`)
var plainResp = []byte("Hello, World!")

func plaintextHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	w.Header().Set("Content-Length", "13")
	w.Write(plainResp)
}

func jsonHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Content-Length", "27")
	w.Write(jsonResp)
}

func echoHandler(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "read error", 500)
		return
	}
	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Length", fmt.Sprintf("%d", len(body)))
	w.Write(body)
}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())

	port := "8080"
	if len(os.Args) > 1 {
		port = os.Args[1]
	}

	fmt.Printf("Go HTTP Server (net/http) on port %s\n", port)

	http.HandleFunc("/plaintext", plaintextHandler)
	http.HandleFunc("/json", jsonHandler)
	http.HandleFunc("/echo", echoHandler)

	if err := http.ListenAndServe("127.0.0.1:"+port, nil); err != nil {
		log.Fatal("ListenAndServe: ", err)
	}
}

File: tests/benchmarks/http/http_server.py
#!/usr/bin/env python3
"""
http_server.py - Python HTTP server (aiohttp) for benchmark.

Usage:
  python3 http_server.py [port]

Default port: 8080
Requires: pip3 install aiohttp
"""

import asyncio
import sys

from aiohttp import web

PLAIN_RESP = b"Hello, World!"
JSON_RESP = b'{"message":"Hello, World!"}'


async def plaintext_handler(request):
    return web.Response(body=PLAIN_RESP, content_type="text/plain")


async def json_handler(request):
    return web.Response(body=JSON_RESP, content_type="application/json")


async def echo_handler(request):
    body = await request.read()
    return web.Response(body=body, content_type="application/octet-stream")


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8080
    print(f"Python HTTP Server (aiohttp) on port {port}")

    app = web.Application()
    app.router.add_get("/plaintext", plaintext_handler)
    app.router.add_get("/json", json_handler)
    app.router.add_post("/echo", echo_handler)

    web.run_app(app, host="127.0.0.1", port=port, print=None,
                access_log=None)


if __name__ == "__main__":
    main()
